### PR TITLE
Credit card brand formatting

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Refactors / DRYs payments code around (re)formatting of credit card brand names.

Came up as part of payments backend work in https://github.com/openfoodfoundation/openfoodnetwork/pull/8503

A significant amount of mess is being thrown out here, and there's a really nice clarification of responsibilities :partying_face:

:fire::fire::fire:

Soundtrack: [Disco Inferno](https://www.youtube.com/watch?v=pG8TyIEAqps) :musical_note: 

#### What should we test?

This PR cleans up some code which translates input for a credit card's brand name (from Stripe for example), which could be in a format like `American Express`, but on our end we need to save it as `american_express`.

I guess we could do a quick check that saving credit cards with Stripe is working correctly? In theory it also applies to anywhere a card can be saved with Stripe (which is a few places), but they have test coverage already.

#### Release notes

Refactored logic around storing credit card brand names in a valid ActiveMerchant format

Changelog Category: Technical changes

### CODE REVIEW

**Review by mkllnk — COMMENTED** [review]
Nice! Thank you.

File: app/models/spree/credit_card.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ class CreditCard < ApplicationRecord
 
     has_many :payments, as: :source
 
+    before_validation :reformat_card_type!
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/models/spree/credit_card.rb]: Ah yes, the setter should be enough on it's own :+1:

File: app/models/spree/credit_card.rb [hunk 2]
```diff
@@ -118,8 +117,8 @@ def save_requested_by_customer?
 
     private
 
-    def reformat_card_type!(type = nil)
-      self[:cc_type] = active_merchant_card_type(type || cc_type)
+    def reformat_card_type!(type = cc_type)
```
Comment (mkllnk) [inline app/models/spree/credit_card.rb:120]: Now we actually don't need the default value any more. :slightly_smiling_face: Doesn't really matter.

**Review by mkllnk — APPROVED** [review]

**Review by mkllnk — APPROVED** [review]

**Review by jibees — APPROVED** [review]
LGTM! 🙏

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Tested saving cards through the `checkout` page and through `/account#/cards` - Is there anywhere else I'm missing? Nothing else comes to mind right now, but do let me know if there is anywhere else I should check...

It all seems to work fine and as before. Below a list of cards created in the context of this PR:

```

  id  | month | year |     cc_type      | last_digits | first_name | last_name | start_month | start_year | issue_number | address_id |         created_at         |         updated_at         | gateway_customer_profile_id |  gateway_payment_profile_id   | user_id | payment_method_id | is_default 
------+-------+------+------------------+-------------+------------+-----------+-------------+------------+--------------+------------+----------------------------+----------------------------+-----------------------------+-------------------------------+---------+-------------------+------------
 4244 | 2     | 2022 | american_express | 8431        |            |           |             |            |              |            | 2021-12-15 19:58:07.426552 | 2021-12-15 19:58:13.169312 | cus_KmcrCDGBcoWQh3          | card_1K73c1FBE7f7kItLhdVqK627 |    6951 |                   | t
 4243 | 2     | 2022 | american_express | 0005        | customer 1 | non CA    |             |            |              |            | 2021-12-15 19:56:13.842105 | 2021-12-15 19:57:28.826174 | cus_KmcpSMG1W9CLrq          | pm_1K73aCFBE7f7kItLVMUTGei9   |    6951 |               756 | f
 4245 | 3     | 2033 | master           | 5100        |            |           |             |            |              |            | 2021-12-15 20:04:21.980199 | 2021-12-15 20:04:21.980199 | cus_KmcyQ7sQ1cg8Ta          | card_1K73i4FBE7f7kItL152Oy7y2 |    6951 |                   | f
 4246 | 3     | 2033 | discover         | 9424        |            |           |             |            |              |            | 2021-12-15 20:05:43.854189 | 2021-12-15 20:05:43.854189 | cus_KmczGBwPPCM9xI          | card_1K73jOFBE7f7kItLeXS7UMs5 |    6951 |                   | f
 4247 | 2     | 2022 | visa             | 0000        | customer 1 | non CA    |             |            |              |            | 2021-12-15 20:06:56.124102 | 2021-12-15 20:06:56.124102 |                             | pm_1K73kZFBE7f7kItLMhlImq24   |    6951 |               756 | f
 4248 | 3     | 2033 | visa             | 0000        | customer 1 | non CA    |             |            |              |            | 2021-12-15 20:09:05.101168 | 2021-12-15 20:09:06.568789 | cus_Kmd2wEPlBUrX2m          | pm_1K73meFBE7f7kItLly8ZmhDh   |    6951 |               756 | f
(6 rows)
```

The only one which is not displayed on the list below is `id=4247`; the reason for this is that this card was indeed used to make a payment, but the Save card checkbox was not ticked. All others appear as expected on the `/account#/cards` page from the customer.

Good to go!!

**Pull request merged**